0_2_pandas_dataframes_attributes_methods.Rmd

```diff
@@ -698,13 +826,153 @@ Data Frame.  To get a sample that is half the size the original Data Frame
+
+```{python}
+# a trick to shuffle the rows
+df.sample(n = len(df))
+```
+
+Or, with greater grace, we can use `frac = 1`, to mean "take a random sample of 100% of the rows", effectively permuting them:
 
 ```{python}
-# Random permutation of rows for entire Data Frame.
+# a clean, efficient way to permute the rows
 df.sample(frac=1)
 ```
 
+If you look at the `index` of either Dataframe in the above ceels, you'll notice that the rows are now in a random order.
+
+Because the sample is a random selection of the rows - and the default setting of `.sample()` is to sample *without replacement*  - using both `.sample(len(df))` and `df.sample(frac=1)` has the effect of randomizing the row order. All the original rows remaining in the sample, and there is no duplication of any rows.
+
+This method is often used on individual columns to (most likely) remove any statistical relationships between the data in that column and other columns. For instance, for [permutation testing](https://matthew-brett.github.io/cfd2020/permutation/permutation_and_t_test.html).
+
+
+::: {solution-end}
+:::
+
+::: {exercise-start}
+:label: more-random
+:class: dropdown
+:::
+
+We mentioned above that the default behaviour of the `.sample()` method is to sample *without replacement*. We also mentioned that you can use the `replace =` argument to sample *with* replacement.
+
+To use a classic "drawing marbles from a bag" analogy, this means that when we randomly select a marble, we place it *back* in the bag (after recording it's identity/colour etc) before we randomly draw the next marble. This means the same marble can appear twice in the final sample. 
+
+As mentioned above, if we draw a sample of rows *with replacement* from our Dataframe, then individual rows can appear more than once in the resultant sample. If we draw a sample with replacement which is *larger* than the original data, it will certainly contain duplicate rows:
+
+```{python}
+# a sample with replacement
+df.sample(len(df) + 10, 
+          replace=True) 
+```
+
+Quickly scan the `index` for duplicates, to see that this it definitely contains duplicates.
+
+You task now is the following:
+
+- take a random sample *with replacement* of all the rows in the Dataframe (e.g. your sample should have the *same number of rows* as the original Dataframe)
+
+- use an appropriate method to calculate the means of `Human Development Index`, `Fertility Rate` and `Population`, from your random sample
+
+- using appropriate Pandas methods and indexing operations, calculate the *difference* between the means of `Human Development Index`, `Fertility Rate` and `Population` in your sample vs the means for the same variables in the original data
+
+- store your result in a Pandas Series (note: depending what methods you use, your result may automatically be stored in a Pandas Series)
+
+The result should look something like this (though your differences with almost certainly be different to the ones shown below):
```

Comment:
```suggestion
The result should look something like this (though your differences will differ from the ones shown below, because of the randomness inherent in your sampling):
```